server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java

```diff
@@ -0,0 +1,363 @@
+                );
+                swapped = clusterRef.compareAndSet(curr, updated);
+            } while (swapped == false);
+        }
+    }
+
+    /**
+     * Executed once when the final reduce is created.
+     *
+     * Note that his will see all the shards, even if they have been passed to the onPartialReduce
+     * method already.
+     *
+     * @param shards The list of shards that are part of this reduce.
+     * @param totalHits The total number of hits in this reduce.
+     * @param aggs The final result for aggregations.
+     * @param reducePhase The version number for this reduce.
+     */
+    @Override
+    public void onFinalReduce(List<SearchShard> shards, TotalHits totalHits, InternalAggregations aggs, int reducePhase) {
+        if (clusters.hasClusterObjects() == false) {
+            return;
+        }
+
+        Map<String, Integer> totalByClusterAlias = shards.stream().collect(Collectors.groupingBy(shard -> {
+            String clusterAlias = shard.clusterAlias();
+            return clusterAlias != null ? clusterAlias : RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY;
+        }, Collectors.reducing(0, e -> 1, Integer::sum)));
+
+        for (Map.Entry<String, Integer> entry : totalByClusterAlias.entrySet()) {
+            String clusterAlias = entry.getKey();
+            int successfulCount = entry.getValue().intValue();
+
+            AtomicReference<SearchResponse.Cluster> clusterRef = clusters.getCluster(clusterAlias);
+            boolean swapped;
+            do {
+                SearchResponse.Cluster curr = clusterRef.get();
+                SearchResponse.Cluster.Status status = curr.getStatus();
+                if (status != SearchResponse.Cluster.Status.RUNNING) {
+                    // don't swap in a new Cluster if the final state has already been set
+                    break;
+                }
+                TimeValue took = new TimeValue(timeProvider.buildTookInMillis());
+                int successfulShards = successfulCount + curr.getSkippedShards();
+                assert successfulShards + curr.getFailedShards() == curr.getTotalShards()
+                    : "successfulShards("
+                        + successfulShards
+                        + ") + failedShards("
+                        + curr.getFailedShards()
+                        + ") != totalShards ("
+                        + curr.getTotalShards()
+                        + ')';
+                if (curr.isTimedOut() || successfulShards < curr.getTotalShards()) {
+                    status = SearchResponse.Cluster.Status.PARTIAL;
+                } else {
+                    assert successfulShards == curr.getTotalShards()
+                        : "successful (" + successfulShards + ") should equal total(" + curr.getTotalShards() + ") if get here";
+                    status = SearchResponse.Cluster.Status.SUCCESSFUL;
+                }
+                SearchResponse.Cluster updated = new SearchResponse.Cluster(
```

Comment:
Maybe add a `with()` copy constructor to `SearchResponse.Cluster` so that we can save at least the first 3 lines 
```
curr.getClusterAlias(),
curr.getIndexExpression(),
curr.isSkipUnavailable(),
```

that are repeated 4 times in this file ?

Reply:
Good idea. I added a SearchResponse.Builder object to only require setting the fields that are changing. Let me know what you think of that impl.